computeIntrinsicSizeAndPreferredAspectRatio should return a value.

sammygill opened the pull request:
#### dcf84a01e958baca2070b539178134241b42a968
<pre>
computeIntrinsicSizeAndPreferredAspectRatio should return a value.
<a href="https://bugs.webkit.org/show_bug.cgi?id=296448">https://bugs.webkit.org/show_bug.cgi?id=296448</a>
<a href="https://rdar.apple.com/156635318">rdar://156635318</a>

Reviewed by Brandon Stewart.

This function passes in two FloatSizes by reference for the callee to
write out to. Instead of doing this, we can just return them by value
since they are pretty small. In this patch we modify the function to
return them as a std::pair&lt;FloatSize, FloatSize&gt;.

There are certainly more cleanups we can do in these individual
functions but I plan to do them in follow ups just to keep each patch
fairly simple.

Canonical link: <a href="https://commits.webkit.org/297864@main">https://commits.webkit.org/297864@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d71bef01a6048d64f9a9b4b9dc979418bfb307e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113088 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32823 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23301 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119295 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63769 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41386 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86073 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41096 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26732 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101727 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66387 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26008 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19858 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63050 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96118 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19931 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122513 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40166 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29958 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94919 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Skipped layout-tests; 71 flakes 80 failures; Uploaded test results; 24 flakes 50 failures; Compiled WebKit; 4 flakes 46 failures; Passed layout tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40550 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97946 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94661 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/24174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39807 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17612 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36291 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40052 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45551 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39693 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43026 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41430 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -454,10 +454,10 @@ void RenderReplaced::computeAspectRatioInformationForRenderBox(RenderBox* conten
     FloatSize intrinsicSize;
     if (shouldApplySizeOrInlineSizeContainment())
-        RenderReplaced::computeIntrinsicSizeAndPreferredAspectRatio(intrinsicSize, preferredAspectRatio);
+        std::tie(intrinsicSize, preferredAspectRatio) = RenderReplaced::computeIntrinsicSizeAndPreferredAspectRatio();
```
In RenderImage.cpp we are using `auto` instead of `std::tie` (line 873). I think it would be good to make this more consistent.

```
auto [intrinsicSize, preferredAspectRatio] = RenderReplaced::computeIntrinsicSizeAndPreferredAspectRatio();
```

sammygill replied:
In the RenderImage case we can use the structured binding syntax to declare and assign the function call return value to the variables. In this case `intrinsicSize` and `preferredAspectRatio` are already declared so we use `std::tie` to assign to them.

stwrt commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
```diff
@@ -93,19 +93,18 @@ FloatSize LegacyRenderSVGRoot::calculateIntrinsicSize() const
     std::optional<FloatSize> intrinsicRatioValue;
+    FloatSize preferredAspectRatio;
     if (!intrinsicSize.isEmpty())
         preferredAspectRatio = { intrinsicSize.width(), intrinsicSize.height() };
```
This isn’t exactly to do with your patch, but it looks like we can set the `preferredAspectRatio` here just to have it overwritten in lines 118-121. Not sure how much that is happening in a real world scenario.

sammygill replied:
Definitely! Much of this code is all over the place so we can probably continue to clean it up

stwrt commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.h:
```diff
@@ -46,7 +46,7 @@ class LegacyRenderSVGRoot final : public RenderReplaced {
     bool isEmbeddedThroughFrameContainingSVGDocument() const;
 
-    void computeIntrinsicSizeAndPreferredAspectRatio(FloatSize& intrinsicSize, FloatSize& preferredAspectRatio) const override;
+    std::pair<FloatSize, FloatSize> computeIntrinsicSizeAndPreferredAspectRatio() const override;
```
Just as a side comment.

When we updated this to return the std::pair, it is a little bit more difficult to identify what we are actually returning and the order without going and reading the function.

sammygill replied:
Yeah I had the same thought as well. I think the other option would be to turn this into a struct, but I ended up deciding to use a `std::pair` because my ultimate goal is to just decouple this function into two separate ones so this should just be a temporary measure

sammygill committed:
computeIntrinsicSizeAndPreferredAspectRatio should return a value.
https://bugs.webkit.org/show_bug.cgi?id=296448
rdar://156635318

Reviewed by Brandon Stewart.

This function passes in two FloatSizes by reference for the callee to
write out to. Instead of doing this, we can just return them by value
since they are pretty small. In this patch we modify the function to
return them as a std::pair<FloatSize, FloatSize>.

There are certainly more cleanups we can do in these individual
functions but I plan to do them in follow ups just to keep each patch
fairly simple.

Canonical link: https://commits.webkit.org/297864@main
